test/forest_test.cpp

```diff
@@ -0,0 +1,290 @@
+
+#if 0 // fullorder_range doesn't exist?
+        BOOST_CHECK(range_value(fullorder_range(f)) == expected);
+#endif
+    }
+
+    /*preorder*/ {
+        auto a = test_edge_traversal<iterator_t, forest_edge::leading>(f, first, last);
+        auto b = test_edge_traversal<const_iterator_t, forest_edge::leading>(f, first, last);
+        BOOST_CHECK(a == b);
+        BOOST_CHECK(to_string(preorder_range(f)) == a);
+    }
+
+    /*postorder*/ {
+        auto a = test_edge_traversal<iterator_t, forest_edge::trailing>(f, first, last);
+        auto b = test_edge_traversal<const_iterator_t, forest_edge::trailing>(f, first, last);
+        BOOST_CHECK(a == b);
+        BOOST_CHECK(to_string(postorder_range(f)) == a);
+    }
+}
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(reverse_traversal) {
+    auto f{big_test_forest()};
+    auto rfirst{std::rbegin(f)};
+    auto rlast{std::rend(f)};
+
+    /*fullorder*/ {
+        static const auto expected{"ABEEDKKJJIIDCHHGGFFCBA"};
+        test_fullorder_traversal<reverse_iterator_t>(rfirst, rlast, expected);
+        test_fullorder_traversal<const_reverse_iterator_t>(rfirst, rlast, expected);
+        BOOST_CHECK(to_string(reverse_fullorder_range(f)) == expected);
+    }
+
+    /*preorder*/ {
+        auto a = test_edge_traversal<reverse_iterator_t, forest_edge::leading>(f, rfirst, rlast);
+        auto b = test_edge_traversal<const_reverse_iterator_t, forest_edge::leading>(f, rfirst, rlast);
+        BOOST_CHECK(a == b);
+    }
+
+    /*postorder*/ {
+        auto a = test_edge_traversal<reverse_iterator_t, forest_edge::trailing>(f, rfirst, rlast);
+        auto b = test_edge_traversal<const_reverse_iterator_t, forest_edge::trailing>(f, rfirst, rlast);
+        BOOST_CHECK(a == b);
+    }
+}
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(child_traversal) {
+    auto f{big_test_forest()};
+    auto parent{std::find_if(f.begin(), f.end(), [](auto& x){ return x == "B"; })};
+    std::string expected;
+
+    BOOST_CHECK(*parent == "B");
+
+    {
+        auto first{child_begin(parent)};
+        auto last{child_end(parent)};
+        expected = to_string(first, last);
+    }
+
+    BOOST_CHECK(to_string(child_range(parent)) == expected);
+
+    if constexpr (false) { // I'm not sure reverse_child_iterator ever worked.
+        forest<std::string>::reverse_child_iterator first{child_begin(parent)};
+        forest<std::string>::reverse_child_iterator last{child_end(parent)};
+        std::string result{to_string(first, last)};
+        BOOST_CHECK(result == expected);
+    }
+}
+
+/**************************************************************************************************/
+
+BOOST_AUTO_TEST_CASE(test_find_parent) {
+    auto f{big_test_forest()};
+
+    {
+        auto child{std::find_if(f.begin(), f.end(), [](auto& x){ return x == "B"; })};
```

Comment:
The lambda appears in the following very often. The code could be reduced if a template functor would be used.

Reply:
Agreed & changed.